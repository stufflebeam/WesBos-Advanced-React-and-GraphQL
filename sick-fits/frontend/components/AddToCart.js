import { faCartPlus } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';
import LoadingAnimation from './LoadingAnimation';
import { useCart } from '../lib/cartState';

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($productId: ID!) {
    addToCart(productId: $productId) {
      id
      quantity
    }
  }
`;

export default function AddToCart({ id }) {
  const cartState = useCart();
  // console.log('[AddToCart] cartState', cartState);
  // const { cartOpen, cartItems, openCart, closeCart } = cartState;
  const { openCart } = cartState;

  // console.log('[AddToCart] id = ', id);
  const [addToCart, { data, loading, error }] = useMutation(
    ADD_TO_CART_MUTATION,
    {
      variables: { productId: id },
      // refetch the currently logged in user to get the updated cart
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  // This is a custom loading animation that I'm using throughout the site.
  // It's a bit more complex than what we built out in the course, but it's
  // also a bit nicer, IMHO. If we'd rather use the 'Add to Cart' -> 'Adding...'
  // approach, then just comment out this conditional, and the rest of the
  // logic is in place already.
  if (loading)
    return (
      <LoadingAnimation
        config={{ height: '12px', width: '12px', color: '#FF0000' }}
      />
    );

  return (
    // <button
    //   type="button"
    //   onClick={(e) => {
    //     console.log('[AddToCart] CLICKED!!! e, id = ', { e, id });
    //     addToCart({ variables: { id } }).catch((err) => console.error(err));
    //   }}
    // >
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        addToCart();
        openCart();
      }}
    >
      Add{loading && 'ing'} to Cart!{' '}
      <FontAwesomeIcon icon={faCartPlus} className="fa-icon" />
    </button>
  );
}
