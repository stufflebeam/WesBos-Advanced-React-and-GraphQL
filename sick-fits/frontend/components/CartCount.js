import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';

const CartCountStyles = styled.div`
  background: var(--red);
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
  /* TODO: Play with the font-size and positioning of the count inside the dot a bit. 
           It seems ever-so-slightly off center. */
  font-size: 1.4rem;

  /* This was the GitHub CoPilot suggested animation code. It seems to pulse consistently. But, I wanted to leave it
     in here so that I can play with it a bit later on. */
  /* 
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 10px;
  font-weight: 100; */
  /* This little bit of magic makes the animation happen */
  /* animation: pulse 0.5s ease-in-out infinite alternate;
  @keyframes pulse {
    from {
      transform: scale(1);
      opacity: 1;
    }
    to {
      transform: scale(1.2);
      opacity: 0.5;
    }
  } */
`;

const AnimationStyles = styled.span`
  position: relative;
  .count {
    position: relative;
    display: block;
    transition: transform 0.4s; /* This needs to match the timing in the CSSTransition's timeout prop */
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <CartCountStyles>{count}</CartCountStyles>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}
