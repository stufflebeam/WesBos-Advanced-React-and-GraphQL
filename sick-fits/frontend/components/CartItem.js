import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  /* align-items: center; */
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

export default function CartItem({ cartItem, id }) {
  const { product } = cartItem;
  if (!product) return null;
  return (
    <CartItemStyles key={id}>
      <img
        width="100"
        src={product.photo.image.publicUrlTransformed}
        alt={product.photo.altText}
      />
      <div>
        <h3>{product?.name}</h3>
        <p>
          {formatMoney(product?.price * cartItem?.quantity)}-
          <em>
            {cartItem?.quantity} &times; {formatMoney(product?.price)}
            each
          </em>
        </p>
      </div>
    </CartItemStyles>
  );
}
