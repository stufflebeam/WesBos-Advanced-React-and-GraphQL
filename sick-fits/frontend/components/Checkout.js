import styled from 'styled-components';
import { loadStripe } from '@stripe/stripe-js';
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from '@stripe/react-stripe-js';
import { useState } from 'react';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/router';
import { useCart } from '../lib/cartState';
import SickButton from './styles/SickButton';
import DisplayError from './ErrorMessage';
import { CURRENT_USER_QUERY } from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;

  // This is styling for any error messages that show up inside the Stripe CardElement
  // see the TODO in the ErrorMessage component for more info about the ideal way to
  // eliminate the need for this sort of custom styling.
  p[data-test='graphql-error'] {
    font-size: 16px !important;
  }
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { closeCart } = useCart();

  // We don't have the token variable yet (at time of definition), so we can't pass it in here
  // Instead, we'll pass it in when we call the function
  const [checkout, { error: graphQLError }] = useMutation(
    CREATE_ORDER_MUTATION,
    {
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  async function handleSubmit(e) {
    // 1. Stop the form from submitting and turn the loader on
    e.preventDefault();
    setLoading(true);
    // console.log('We gotta do some work...');
    // 2. Start the page transition
    nProgress.start();
    // 3. Create the payment method via Stripe (Token comes back here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });
    //   .then((res) => {
    //     console.log(res);
    //   })
    //   .catch((err) => {
    //     setError(err);
    //   });
    console.log('[Checkout]', { error, paymentMethod });
    // 4. Handle any errors from Stripe
    if (error) {
      setError(error);
      nProgress.done();
      return; // stops the checkout from happening
    }
    // 5. Send the token from step 3 to our Keystone server, via a custom mutation!
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });
    console.log('[Checkout] Finished with the order:', order);

    // 6. Change the page to view the order

    router.push({
      pathname: `/order/[id]`,
      query: { id: order.data.checkout.id },
      // The following (original) syntax worked as expected for me, but I am switching to the version
      // used in the video (for now), in order to keep things consistent with the rest of the course code.
      // query: { id: order.data.checkout.id },
    });

    // 7. Close the cart (if the payment was successful)

    closeCart();

    // 8. Turn the loader off
    setLoading(false);
    nProgress.done();
  }

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <DisplayError error={error} />}
      {graphQLError && <DisplayError error={graphQLError} />}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
}

function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}

export { Checkout };
