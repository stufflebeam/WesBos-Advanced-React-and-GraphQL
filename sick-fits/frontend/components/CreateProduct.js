import useForm from '../lib/useForm';

export default function CreateProduct() {
  const placeholder =
    'https://res.cloudinary.com/stufflebeam/image/upload/v1661582104/placeholder/placeholder-images-product-2_large.webp';

  const { inputs, handleChange, resetForm, clearForm } = useForm({
    name: 'Test Product',
    description: 'This is a test product',
    price: 929,
    image: placeholder,
  });

  return (
    <form>
      <label>
        Name:
        <input
          type="text"
          name="name"
          placeholder="Name"
          value={inputs.name}
          onChange={handleChange}
        />
      </label>
      <label>
        Description:
        <input
          type="text"
          name="description"
          placeholder="Description"
          value={inputs.description}
          onChange={handleChange}
        />
      </label>
      <label>
        Price:
        <input
          type="number"
          name="price"
          placeholder="Price"
          value={inputs.price}
          onChange={handleChange}
        />
      </label>
      <img src={inputs.image} alt={inputs.name} />
      <button type="button" onClick={clearForm}>
        clear
      </button>
      <button type="button" onClick={resetForm}>
        reset
      </button>
    </form>
  );
}
