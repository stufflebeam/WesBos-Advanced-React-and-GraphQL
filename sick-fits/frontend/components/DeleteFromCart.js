import { faTrash } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import styled from 'styled-components';
import { CURRENT_USER_QUERY } from './User';
import LoadingAnimation from './LoadingAnimation';
// import { useCart } from '../lib/cartState';

const BigButton = styled.button`
  font-size: 2rem;
  background: none;
  border: 0;
  &:hover {
    color: var(--red);
    cursor: pointer;
  }
`;

const DELETE_FROM_CART_MUTATION = gql`
  mutation DELETE_FROM_CART_MUTATION($id: ID!) {
    deleteCartItem(id: $id) {
      id
      # quantity
      # product {
      #   id
      #   name
      #   price
      #   description
      # }
      # user {
      #   id
      #   name
      #   email
      # }
    }
  }
`;
function update(cache, payload) {
  cache.evict(cache.identify(payload.data.deleteCartItem));
}

export default function DeleteFromCart({ id }) {
  // const cartState = useCart();
  // console.log('[DeleteFromCart] cartState', cartState);
  // const { cartOpen, cartItems, openCart, closeCart } = cartState;
  // const { openCart } = cartState;

  // console.log('DeleteFromCart] id = ', id);
  const [deleteFromCart, { data, loading, error }] = useMutation(
    DELETE_FROM_CART_MUTATION,
    {
      variables: { id },
      // refetch the currently logged in user to get the updated cart
      // refetchQueries: [{ query: CURRENT_USER_QUERY }],
      // Instead of going back to the network to refetch the user query, just evict the item from the cache
      update,
      // Having to comment this out for now, because it's currently causing an error
      // This appears to be a bug in Keystone. So, just waiting for a fix before re-enabling this optimistic response.
      // optimisticResponse: {
      //   __typename: 'Mutation',
      //   deleteCartItem: {
      //     __typename: 'CartItem',
      //     id,
      //   },
      // },
    }
  );

  // This is a custom loading animation that I'm using throughout the site.
  // It's a bit more complex than what we built out in the course, but it's
  // also a bit nicer, IMHO. If we'd rather use the 'Add to Cart' -> 'Adding...'
  // approach, then just comment out this conditional, and the rest of the
  // logic is in place already.
  if (loading)
    return (
      <LoadingAnimation
        config={{ height: '12px', width: '12px', color: '#FF0000' }}
      />
    );

  return (
    <BigButton
      type="button"
      disabled={loading}
      onClick={deleteFromCart}
      title="Remove this item from the cart"
    >
      <FontAwesomeIcon icon={faTrash} className="fa-icon" />
    </BigButton>
  );
}
