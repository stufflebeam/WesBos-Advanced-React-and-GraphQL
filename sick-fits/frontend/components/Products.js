import gql from 'graphql-tag';
// auto-import returned a much deeper link (see below), but the above
// import is the right one, I believe. So, that's what I'm going with.
// import gql from "../node_modules/graphql-tag/lib/index";
// auto-import is having trouble importing useQuery as well now...
// unsure what's going on here. :(
import { useQuery } from '@apollo/client';
import ReactLoading from 'react-loading';
import styled from 'styled-components';
import LoadingAnimation from './LoadingAnimation';
import Product from './Product';

export const ALL_PRODUCTS_QUERY = gql`
  query ALL_PRODUCTS_QUERY {
    allProducts {
      id
      name
      price
      description
      photo {
        id
        altText
        image {
          publicUrlTransformed
        }
      }
    }
  }
`;

const ProductsListStyles = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  /* max-width: ${(props) => props.theme.maxWidth};
  margin: 0 auto; */
`;

export default function Products() {
  const { data, error, loading } = useQuery(ALL_PRODUCTS_QUERY);
  //   console.log('[Products]: data:', data);
  //   console.log('[Products]: error:', error);
  //   console.log('[Products]: loading:', loading);

  if (loading) return <LoadingAnimation config={{ color: '#FF0000' }} />;
  if (error) return <p>Error: {error.message}</p>;
  return (
    <div>
      <ProductsListStyles>
        {data.allProducts.map((product) => (
          <Product key={product.id} product={product} />
        ))}
      </ProductsListStyles>
    </div>
  );
}
