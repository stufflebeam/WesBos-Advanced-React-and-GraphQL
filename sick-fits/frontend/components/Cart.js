import { faShoppingCart, faXmark } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import CartItem from './CartItem';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import { useUser } from './User';
import { useCart } from '../lib/cartState';

const CloseCartStyles = styled.span`
  font-size: 4rem;
  right: 3rem;
  position: absolute;
  :hover {
    color: red;
    cursor: pointer;
  }
`;

export default function Cart() {
  const currentUser = useUser();
  console.log('[Cart] currentUser', currentUser);
  if (!currentUser) {
    return null;
  }

  const cartState = useCart();
  console.log('[Cart] cartState', cartState);
  const { cartOpen, cartItems, closeCart } = cartState;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>
          {currentUser?.name}'s Cart{' '}
          <FontAwesomeIcon icon={faShoppingCart} className="fa-icon" />
        </Supreme>
        <CloseCartStyles type="button" onClick={closeCart}>
          <FontAwesomeIcon icon={faXmark} className="fa-icon" />
        </CloseCartStyles>
      </header>
      <ul>
        {currentUser?.cart.map((cartItem) => (
          <CartItem key={cartItem.id} id={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(currentUser.cart))}</p>
      </footer>
    </CartStyles>
  );
}
