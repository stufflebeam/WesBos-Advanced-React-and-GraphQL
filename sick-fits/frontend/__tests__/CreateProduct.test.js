import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import Router from 'next/router'; // We will mock this
import wait from 'waait';
import { waitFor } from '@testing-library/dom';
import { act } from 'react-dom/test-utils';
import CreateProduct, {
  CREATE_PRODUCT_MUTATION,
} from '../components/CreateProduct';
import { ALL_PRODUCTS_QUERY } from '../components/Products';
import { fakeItem, makePaginationMocksFor } from '../lib/testUtils';

const item = fakeItem();

jest.mock('next/router', () => ({
  push: jest.fn(),
}));

const mocks = [
  // Create Product Mutation Mock
  {
    request: {
      query: CREATE_PRODUCT_MUTATION,
      variables: {
        name: item.name,
        description: item.description,
        price: item.price,
        image: '',
      },
    },
    result: {
      data: {
        createProduct: {
          __typename: 'Item',
          ...item, // Spread in all the fake item fields
          //   id: 'abc123',
          //   name: item.name,
          //   description: item.description,
          //   price: item.price,
          //   image: {},
        },
      },
    },
  },
  // All Products Query Mock
  {
    request: { query: ALL_PRODUCTS_QUERY, variables: { skip: 0, first: 2 } },
    result: {
      data: {
        allProducts: [item],
      },
    },
  },
];

describe('<CreateProduct />', () => {
  it('renders and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    // debug(container);
    expect(container).toMatchSnapshot();
  });

  it('handles the updating of form elements', async () => {
    // 1. Render the form
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    // 2. Type into the form fields
    // await userEvent.type(screen.getByPlaceholderText(/name/i), item.name);
    await userEvent.type(screen.getByPlaceholderText('Name'), item.name);
    // await userEvent.type( screen.getByPlaceholderText(/description/i), item.description );
    await userEvent.type(
      screen.getByPlaceholderText('Description'),
      item.description
    );
    // await userEvent.type(
    //   screen.getByPlaceholderText(/price/i),
    //   item.price.toString()
    // );
    // await userEvent.type(
    //   screen.getByTestId('create-product-price'),
    //   item.price.toString()
    // );
    const priceInput = screen.getByTestId('create-product-price');
    expect(priceInput).toBeInTheDocument();
    // NOTE: We must clear() the value of the input box before typing in the new value
    //       as otherwise the value will be appended to the existing value, so typing
    //       in '5000' would result in a display value of '05000' instead of '5000'
    //       See this Slack post for more details:
    //       https://wesbos.slack.com/archives/C9G96G2UB/p1642598592183300?thread_ts=1642560686.178400&cid=C9G96G2UB
    userEvent.clear(priceInput);
    await userEvent.type(priceInput, item.price.toString());

    // 3. Check that the inputs have the expected values
    // debug(container);
    expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.price)).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("creates a new product and navigates to the new product's page when the form is submitted", async () => {
    // NOTE: Could also create the mocks here inside this test, if they are only needed here.
    //       However, I prefer to keep the mocks outside of the test, so that they are
    //       accessible to other tests if I end up wanting to use them elsewhere.
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <CreateProduct />
      </MockedProvider>
    );
    // debug(container);
    //   expect(container).toMatchSnapshot();
    // 1. Type into the form fields
    await userEvent.type(screen.getByPlaceholderText('Name'), item.name);
    await userEvent.type(
      screen.getByPlaceholderText('Description'),
      item.description
    );
    const priceInput = screen.getByTestId('create-product-price');
    expect(priceInput).toBeInTheDocument();
    // NOTE: We must clear() the value of the input box before typing in the new value (SEE ABOVE FOR MORE INFORMATION)
    userEvent.clear(priceInput);
    await userEvent.type(priceInput, item.price.toString());
    const addProductButton = screen.getByRole('button', {
      name: /Add Product/i,
    });
    // debug(addProductButton);

    // FIXME: Added this explicit call to act() to try an get around the warnings in the console,
    //        but it's still triggering them.
    //        The underlying issue is that the form is cleared during the try block of the product
    //        creation process. Inside useForm, there is a function called clearForm, which is
    //        which is called by CreateProduct. This, I believe, is causing the warnings by updating
    //        the DOM with the cleared form after our tests are done. However, I've tested for these
    //        In several spots here, with a number of timeout settings, and the warnings persist.
    //        Even with the clearForm call commented out, I'm still getting the warnings though. So,
    //        I'm not sure exactly what's going on here. Need to revist this to fix.
    await act(async () => await userEvent.click(addProductButton));
    // await userEvent.click(addProductButton);

    await waitFor(() => wait(0));
    // Check that the form inputs have been reset
    expect(screen.getByPlaceholderText('Name')).toHaveValue('');
    expect(screen.getByPlaceholderText('Description')).toHaveValue('');
    expect(priceInput).toHaveValue(null);

    await waitFor(() => wait(40));

    // Wait to see if the page change is initiated (and happens successfully)
    expect(Router.push).toHaveBeenCalled();
    expect(Router.push).toHaveBeenCalledWith({
      pathname: `/product/${item.id}`,
    });

    // const productComponent = await screen.findByTestId('product-component');
    // debug();
  });
});
