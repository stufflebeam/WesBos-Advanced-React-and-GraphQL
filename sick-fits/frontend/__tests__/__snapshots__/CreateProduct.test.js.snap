// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct /> handles the updating of form elements 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image:
        <input
          name="image"
          required=""
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name:
        <input
          name="name"
          placeholder="Name"
          type="text"
          value="dogs are best"
        />
      </label>
      <label
        for="description"
      >
        Description:
        <textarea
          id="description"
          name="description"
          placeholder="Description"
        >
          dogs
        </textarea>
      </label>
      <label
        for="price"
      >
        Price:
        <input
          data-testid="create-product-price"
          name="price"
          placeholder="Price"
          type="number"
          value="0"
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;

exports[`<CreateProduct /> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 hVEEdl"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="image"
      >
        Image:
        <input
          name="image"
          required=""
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name:
        <input
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
      </label>
      <label
        for="description"
      >
        Description:
        <textarea
          id="description"
          name="description"
          placeholder="Description"
        />
      </label>
      <label
        for="price"
      >
        Price:
        <input
          data-testid="create-product-price"
          name="price"
          placeholder="Price"
          type="number"
          value="0"
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
